src/System.Private.ServiceModel/src/System/ServiceModel/Channels/MessageContent.cs

```diff
@@ -0,0 +1,266 @@
+                            if (actionParams.Length > 0)
+                            {
+                                try
+                                {
+                                    string value = string.Format(CultureInfo.InvariantCulture, "\"{0}\"", actionParams[0].Value);
+                                    if (string.Compare(value, action, StringComparison.Ordinal) != 0)
+                                    {
+                                        throw DiagnosticUtility.ExceptionUtility.ThrowHelperError(
+                                            new ProtocolException(SR.Format(SR.HttpSoapActionMismatchContentType, action, value)));
+                                    }
+                                    shouldSetContentType = false;
+                                }
+                                catch (FormatException formatException)
+                                {
+                                    throw DiagnosticUtility.ExceptionUtility.ThrowHelperError(
+                                        new ProtocolException(SR.Format(SR.HttpContentTypeFormatException, formatException.Message, Headers.ContentType.ToString()), formatException));
+                                }
+                            }
+                        }
+
+                        if (shouldSetContentType)
+                        {
+                            Headers.ContentType.Parameters.Add(new NameValueHeaderValue("action", action));
+                        }
+                    }
+                }
+            }
+        }
+
+        private bool SetContentType(string contentType)
+        {
+            MediaTypeHeaderValue contentTypeHeaderValue;
+            if (MediaTypeHeaderValue.TryParse(contentType, out contentTypeHeaderValue))
+            {
+                Headers.ContentType = contentTypeHeaderValue;
+                return true;
+            }
+            return false;
+        }
+
+        protected override void Dispose(bool disposing)
+        {
+            if(disposing && !_disposed)
+            {
+                _disposed = true;
+                if (_stream != null)
+                {
+                    var stream = _stream;
+                    _stream = null;
+                    stream.Dispose();
+                }
+            }
+
+            base.Dispose(disposing);
+        }
+
+        internal static HttpContent Create(HttpChannelFactory<IRequestChannel> factory, Message request, TimeoutHelper _timeoutHelper)
+        {
+            if(TransferModeHelper.IsRequestStreamed(factory.TransferMode))
+            {
+                return new StreamedMessageContent(request, factory.MessageEncoderFactory.Encoder);
+            }
+            else
+            {
+                return new BufferedMessageContent(request, factory.MessageEncoderFactory.Encoder, factory.BufferManager);
+            }
+        }
+    }
+
+    internal class StreamedMessageContent : MessageContent
+    {
```

Comment:
I'd consider making these classes public solely for unit testing.  If would be extremely useful to unit test these